src/token/src/canister.rs

```diff
@@ -275,9 +275,17 @@ impl TokenCanister {
         mint(self, to, amount)
     }
 
+    /// Burn `amount` of tokens from `from` principal.
+    /// If `from` is None, then caller's tokens will be burned.
+    /// If `from` is Some(_) but method called not by owner, `TxError::Unauthorized` will be returned.
+    /// If owner calls this method and `from` is Some(who), then who's tokens will be burned.
     #[update]
-    fn burn(&self, amount: Nat) -> TxReceipt {
-        burn(self, amount)
+    fn burn(&self, from: Option<Principal>, amount: Nat) -> TxReceipt {
+        if from.is_some() {
```

Comment:
Nice one Kirill, we probably want to do the check inside inspect message. Please see: https://smartcontracts.org/docs/interface-spec/index.html#system-api-inspect-message

Reply:
We need the check both here and in inspect message, because inter-canister calls are not inspected. Inspect message has the check at the moment, but it needs to be modified to take the `from` argument into account.